File: scripts/vision-ocr.js
#!/usr/bin/env osascript -l JavaScript
// vision-ocr.js — Apple Vision Framework OCR via JXA
// Usage: osascript -l JavaScript vision-ocr.js <image-path> [lang1,lang2]
// Output: JSON on stdout

ObjC.import('Foundation')
ObjC.import('AppKit')
ObjC.import('Vision')

function run(argv) {
    if (argv.length < 1) {
        return JSON.stringify({ error: "Usage: vision-ocr.js <image-path> [langs]" })
    }

    const imagePath = argv[0]
    const languages = argv.length >= 2 ? argv[1].split(',') : []

    const url = $.NSURL.fileURLWithPath(imagePath)
    const image = $.NSImage.alloc.initWithContentsOfURL(url)

    if (!image) {
        return JSON.stringify({ error: "Cannot load image: " + imagePath })
    }

    const cgImage = image.CGImageForProposedRectContextHints($.nil, $.nil, $.nil)

    const request = $.VNRecognizeTextRequest.alloc.init
    request.recognitionLevel = $.VNRequestTextRecognitionLevelAccurate
    request.usesLanguageCorrection = true

    if (languages.length > 0) {
        request.recognitionLanguages = $.NSArray.arrayWithArray(languages)
    }

    const handler = $.VNImageRequestHandler.alloc.initWithCGImageOptions(cgImage, $.NSDictionary.alloc.init)
    handler.performRequestsError($([request]), $())

    const results = request.results
    const count = results.count
    let text = ''
    let totalConfidence = 0.0

    for (let i = 0; i < count; i++) {
        const obs = results.objectAtIndex(i)
        const candidate = obs.topCandidates(1).objectAtIndex(0)
        text += (text === '' ? '' : '\n') + candidate.string.js
        totalConfidence += candidate.confidence
    }

    return JSON.stringify({
        text: text,
        confidence: count > 0 ? totalConfidence / count : 0,
        lineCount: count
    })
}

File: scripts/vision-ocr.swift
#!/usr/bin/env swift
// vision-ocr.swift — Apple Vision Framework OCR CLI
// Usage: vision-ocr.swift <image-path> [language]
// Outputs JSON: {"text": "...", "confidence": 0.95, "observations": [...]}

import Foundation
import Vision
#if canImport(AppKit)
import AppKit
#endif

func ocrImage(at path: String, languages: [String]) throws -> [String: Any] {
    let url = URL(fileURLWithPath: path)
    guard FileManager.default.fileExists(atPath: path) else {
        throw NSError(domain: "vision-ocr", code: 1, userInfo: [NSLocalizedDescriptionKey: "File not found: \(path)"])
    }

    #if canImport(AppKit)
    guard let image = NSImage(contentsOf: url) else {
        throw NSError(domain: "vision-ocr", code: 2, userInfo: [NSLocalizedDescriptionKey: "Cannot load image: \(path)"])
    }
    guard let cgImage = image.cgImage(forProposedRect: nil, context: nil, hints: nil) else {
        throw NSError(domain: "vision-ocr", code: 3, userInfo: [NSLocalizedDescriptionKey: "Cannot create CGImage"])
    }
    #else
    throw NSError(domain: "vision-ocr", code: 4, userInfo: [NSLocalizedDescriptionKey: "AppKit not available"])
    #endif

    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.recognitionLanguages = languages.isEmpty ? ["en-US"] : languages
    request.usesLanguageCorrection = true
    // Set revision for best accuracy
    if #available(macOS 13.0, *) {
        request.revision = VNRecognizeTextRequestRevision3
    }

    let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
    try handler.perform([request])

    guard let results = request.results else {
        return ["text": "", "confidence": 0.0, "observations": []]
    }

    var fullText = ""
    var totalConfidence: Float = 0.0
    var observations: [[String: Any]] = []
    var lineCount = 0

    for observation in results {
        guard let topCandidate = observation.topCandidates(1).first else { continue }

        let text = topCandidate.string
        let confidence = topCandidate.confidence

        let bbox = observation.boundingBox
        let bboxDict: [String: Double] = [
            "x": Double(bbox.origin.x),
            "y": Double(bbox.origin.y),
            "width": Double(bbox.size.width),
            "height": Double(bbox.size.height)
        ]

        observations.append([
            "text": text,
            "confidence": Double(confidence),
            "bbox": bboxDict
        ])

        fullText += (fullText.isEmpty ? "" : "\n") + text
        totalConfidence += confidence
        lineCount += 1
    }

    let avgConfidence = lineCount > 0 ? Double(totalConfidence) / Double(lineCount) : 0.0

    return [
        "text": fullText,
        "confidence": avgConfidence,
        "observations": observations
    ]
}

// Main
let args = CommandLine.arguments
guard args.count >= 2 else {
    let err: [String: Any] = ["error": "Usage: vision-ocr.swift <image-path> [lang1,lang2,...]"]
    let data = try! JSONSerialization.data(withJSONObject: err, options: .prettyPrinted)
    FileHandle.standardError.write(data)
    exit(1)
}

let imagePath = args[1]
let languages: [String] = args.count >= 3 ? args[2].split(separator: ",").map(String.init) : []

do {
    let result = try ocrImage(at: imagePath, languages: languages)
    let data = try JSONSerialization.data(withJSONObject: result, options: .prettyPrinted)
    FileHandle.standardOutput.write(data)
} catch {
    let err: [String: Any] = ["error": error.localizedDescription]
    let data = try! JSONSerialization.data(withJSONObject: err, options: .prettyPrinted)
    FileHandle.standardError.write(data)
    exit(1)
}
